cpp/src/arrow/record_batch.h

```diff
@@ -254,6 +255,16 @@ class ARROW_EXPORT RecordBatch {
   /// \return Status
   virtual Status ValidateFull() const;
 
+  /// \brief Return a top-level sync event object for this record batch
+  ///
+  /// If all of the data for this record batch is in host memory, then this
+  /// should return null (the default impl). If the data for this batch is
+  /// on a device, then if synchronization is needed before accessing the
+  /// data the returned sync event will allow for it.
```

Comment:
I don't fully understand this part. AFAIK a RecordBatch currently is agnostic the device of its buffers. So for example you can have a RecordBatch backed by buffers that live in CUDA memory, but then this method will always hardcoded return NULL, which is not correct in that case?

Reply:
Yea, at the moment I don't have an implementation there. I put a note on the `ImportDeviceRecordBatchReader` documentation if you look at bridge.h:

```
/// We are not yet bubbling the sync events from the buffers up to
/// the `GetSyncEvent` method of an imported RecordBatch. This will be added in a future
/// update.
```

Reply:
Is the idea that longer term there would be a generic implementation here in RecordBatch that checks the sync events of its underlying buffers? Because in practice, we don't subclass RecordBatch in Arrow C++ to add CUDA support, so that method cannot be overriden

Reply:
Okay, I've propagated the event and device_type throughout the record batch and the `Make` functions (defaulting it to nullptr) which should allow us to ensure this is all correct now. I'll update the documentation comments accordingly